### PR TITLE
8280798: com.sun.jdi.ObjectReference::setValue spec should prohibit any final field modification

### PR DESCRIPTION
**Pull request by alexmenkov**
com.sun.jdi.ObjectReference::setValue spec says that final static fields cannot be modified, but openjdk implementation throws IllegalArgumentException for any final fields (static or instance).

The fix updates the spec to prohibit any final field modification
CSR: JDK-8281652

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change requires a CSR request to be approved

### Issues
 * [JDK-8280798](https://bugs.openjdk.org/browse/JDK-8280798): com.sun.jdi.ObjectReference::setValue spec should prohibit any final field modification
 * [JDK-8281652](https://bugs.openjdk.org/browse/JDK-8281652): com.sun.jdi.ObjectReference::setValue spec should prohibit any final field modification (**CSR**)

### Reviewers
 * [Alan Bateman](https://openjdk.org/census#alanb) (@AlanBateman - **Reviewer**) ⚠️ Review applies to [5d98874c](https://git.openjdk.org/jdk/pull/11279/files/5d98874cbea6675fa352a6d271ce953150f54f06)
 * [Chris Plummer](https://openjdk.org/census#cjplummer) (@plummercj - **Reviewer**)
 * [Serguei Spitsyn](https://openjdk.org/census#sspitsyn) (@sspitsyn - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk pull/11279/head:pull/11279` \
`$ git checkout pull/11279`

Update a local copy of the PR: \
`$ git checkout pull/11279` \
`$ git pull https://git.openjdk.org/jdk pull/11279/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 11279`

View PR using the GUI difftool: \
`$ git pr show -t 11279`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/11279.diff">https://git.openjdk.org/jdk/pull/11279.diff</a>

</details>

### CODE REVIEW

**Comment by sspitsyn** [comment]
This issue need a release note.
The fix itself looks good.

**Comment by dholmes-ora** [comment]
I commented in the bug report, but shouldn't a debugger be able to set a final field?

**Comment by AlanBateman** [comment]
I think this one is going to require more investigation as I can't immediately tell if this is a spec or implementation issue.

I checked the JDWP spec and the ClassReference/SetValues has "Final fields cannot be set" in its spec, so this means that static finals can't be set. ObjectReference/SetValues command doesn't have this, so JDWP allows a tool to set final fields. So if the change proposed in this PR goes ahead it means that JDI and JDWP specs won't be aligned (which is okay, it just needs to be fully understood).

Maybe start by digging into the history of ReferenceTypeImpl to see if the spec vs. implementation discrepancy dates goes back to JDK 1.2 (when JPDA was added) or it crept in at some other time.  Maybe also dig into bug reports to see if there has been any complaints about the current behavior.

If the outcome is that spec is changed as proposed then it will require the description for IllegalArgumentException to be expanded. If the outcome is to allow for changing of some final fields then the spec may need to be updated to take account of "non-modifiable final field", as in final fields in hidden classes and records. In both cases, the JDWP spec will need to looked at it too.

**Comment by plummercj** [comment]
ObjectReference.setValue() calls ReferenceType.validateFieldSet(), which does:

    void validateFieldSet(Field field) {
        validateFieldAccess(field);
        if (field.isFinal()) {
            throw new IllegalArgumentException("Cannot set value of final field");
        }
    }

This code has been in place since the file was introduced on 2007-12-01. If there is a version that predates it, I don't know where to look. I'm not sure of the history relative to JDK 1.2 and the introduction of JPDA.

Yes, what is being proposed makes the JDI and JDWP specs out of sync, but the implementations have always been out of sync. The proposed spec change is just clarifying what the implementation already does. However, Eclipse does allow the user to set a non-static final field, but only after clicking through a warning dialog.

**Comment by dholmes-ora** [comment]
So changing the spec would make Eclipse non-compliant. Perhaps to bring the spec and implementation in-line we have to make the implementation behaviour allowable by the spec but not required - that is what we often do in other areas (e.g. JNI spec - my personal point of experience).

But I remain surprised/perplexed that a debugger would not be allowed to change a final field. Can't we still change final fields via reflection anyway? (I know the constraints have been tightening in that area.)

**Comment by plummercj** [comment]
https://docs.oracle.com/en/java/javase/18/docs/api/java.base/java/lang/reflect/Field.html#set(java.lang.Object,java.lang.Object)

It is allowed, for the most part, but with the warning: 

`
Setting a final field in this way is meaningful only during deserialization or reconstruction of instances of classes with blank final fields, before they are made available for access by other parts of a program. Use in any other context may have unpredictable effects, including cases in which other parts of a program continue to use the original value of this field.`

In [JDK-8280798](https://bugs.openjdk.org/browse/JDK-8280798) Alex has documented a use case that shows the dangers of allowing this:

https://bugs.openjdk.org/browse/JDK-8280798?focusedCommentId=14539651&page=com.atlassian.jira.plugin.system.issuetabpanels:comment-tabpanel#comment-14539651

**Comment by alexmenkov** [comment]
The change makes Eclipse non-compliant, but it shouldn't be a problem as Eclipse JDI implementation is used only by Eclipse debugger (we discussed it with one of the Eclipse debugger contributors - he doesn't see problem here). And it's up to Eclipse team to decide if they want to update the implementation or keep it.
Relaxing the spec was one of the solutions considered, but looks like it doesn't make much sense.

Reflection allows to modify final instance fields, but its spec has a warning that changing final fields is meaningful only in certain cases (which are not common for debuggers). So if we decide to update implementation to modify final fields, it would be better to add similar warning to JDI spec (additionally to "non-modifiable final field" changes described by Alan)

**Comment by plummercj** [comment]
I also have to wonder about JIT behavior in this context (and this might be what the RMI warning is alluding to). What if the JIT sees a static final reference to an object with a final field. It probably generates code that references the value of the field rather than generating code that fetches from the field. If the field value changes, the JIT won't know and the existing code will continue to reference the old value.

**Comment by AlanBateman** [comment]
I dug into the pre-OpenJDK history as most of the action on this was in 1999.

In 1999,  ObjectReference.setValue's docs changed from "The field must not be final" to "If static, the field must not be final". The exception message started out as "Cannot set value of final field" and changed to "Cannot set value of static final field".  

jjh re-visited it in 2005 via JDK-623152 where he changed the exception message to "Cannot set value of final field".

I don't think the ancient history sheds much more light on the original intent. I suspect the original intent was to disallow setValue only static final fields. That would have aligned with Field.set + setAccessible(true) at the time.

In general I think the JDK is on the path to a world where "final means final". We see this with final fields on hidden classes and fields in records. So the proposal to change the JDI spec may be too bad but it require reaching out to IDE maintainers. Eclipse have their own JDI implementation. VS.Code uses the Eclipse JDI implementation. IntelliJ maintain a fork of JDI. So while the JDK's implementation of JDI has disallow setting of final fields, it seems that at least two of main stream IDEs do allow it because they are using the Eclipse implementation.

**Comment by alexmenkov** [comment]
I decided to check if VS Code shows the warning as Eclipse does, so I installed VS Code and tried to set final fields in debugger.
For both static and instance fields I got:
Failed to set variable. Reason: java.lang.UnsupportedOperationException: SetVariableRequest: Final field value cannot be changed.

So VS Code DOES NOT allow to set static final fields.

As far as I got VS Code uses Eclipse JDT only for editor (code navigation, auto completion, refactoring, etc), but debugging is implemented by another extension: Debugger for Java (https://github.com/Microsoft/java-debug). I.e. VS Code does not use Eclipse 
JDI implementation.

**Comment by plummercj** [comment]
> IntelliJ currently does not allow setting final fields. I'm checking with a team member to make sure they are ok with the spec change.

Egor at Jetbrains (Intellij) is ok with the proposed spec changes.

**Comment by AlanBateman** [comment]
I checked with the Microsoft team that maintain the Java Debugger in VS Code. The JDI implementation comes from the Eclipse JDT project. The summary is that the VS Code debugger doesn't allow final fields to be modified. It checks if the field is final so it won't call ObjectReference::setValue for final fields.

**Review by AlanBateman — APPROVED** [review]

**Comment by AlanBateman** [comment]
I think we've got enough information now to support with this change.

**Comment by AlanBateman** [comment]
One other thing to add is that we should survey the existing tests to make sure that there are tests to ensure IAE is thrown for both static and instance fields when they are final.

**Comment by plummercj** [comment]
> One other thing to add is that we should survey the existing tests to make sure that there are tests to ensure IAE is thrown for both static and instance fields when they are final.

I think a good way to find such tests would be to remove the !isFinal() check and see if any tests fail.

**Comment by alexmenkov** [comment]
The only test for final fields is vmTestbase/nsk/jdi/ObjectReference/setValue/setvalue004.
It tested only static final fields.
I added testing for instance final fields

File: test/hotspot/jtreg/vmTestbase/nsk/jdi/ObjectReference/setValue/setvalue004t.java [hunk 1]
```diff
@@ -42,6 +42,17 @@ public class setvalue004t {
...
+    // tested static final fields
+    static final byte    iByteFld = 127;
+    static final short   iShortFld = -32768;
+    static final int     iIntFld = 2147483647;
+    static final long    iLongFld = 9223372036854775807L;
+    static final float   iFloatFld = 5.1F;
+    static final double  iDoubleFld = 6.2D;
+    static final char    iCharFld = 'a';
+    static final boolean iBooleanFld = false;
+    static final String  iStrFld = "instance field";
```
Comment (plummercj) [inline test/hotspot/jtreg/vmTestbase/nsk/jdi/ObjectReference/setValue/setvalue004t.java:54]: I don't think you meant for these fields to be static.

Comment (alexmenkov) [reply test/hotspot/jtreg/vmTestbase/nsk/jdi/ObjectReference/setValue/setvalue004t.java]: Oh, sure. Incomplete change (or I did undo accidentally)

Comment (alexmenkov) [reply test/hotspot/jtreg/vmTestbase/nsk/jdi/ObjectReference/setValue/setvalue004t.java]: Fixed.

**Review by plummercj — APPROVED** [review]

**Review by sspitsyn — APPROVED** [review]

**Pull request closed without merging**